keysign/app.py

```diff
@@ -171,61 +174,54 @@ def on_activate(self, app):
-
+        reactor.run()
 
     def run(self, args=[]):
         super(KeysignApp, self).run()
 
-    def on_key_activated(self, widget, key):
-        log.info("Activated key %r", key)
-        # Ouf, we rely on the the SendApp to have reacted to
-        # the signal first, so that it sets up the keypresentwidget
-        # and so that we can access it here.  If it did, however,
-        # We might not be able to catch the mapped signal quickly
-        # enough. So we ask the widget wether it is already mapped.
-        kpw = self.send.kpw
-        kpw.connect('map', self.on_keypresent_mapped)
-        log.debug("KPW to wait for map: %r (%r)", kpw, kpw.get_mapped())
-        if kpw.get_mapped():
-            # The widget is already visible. Let's quickly call our handler
-            self.on_keypresent_mapped(kpw)
-
-        ####
-        # Saving subtitle
-        self.headerbar_subtitle = self.headerbar.get_subtitle()
-        self.headerbar.set_subtitle("Sending {}".format(key.fpr))
-        ####
-        # Making button clickable
-        self.header_button.set_sensitive(True)
-
-
-
-
+    @staticmethod
+    def on_delete_window(*args):
+        reactor.callFromThread(reactor.stop)
+        Gtk.main_quit(*args)
 
     def on_sr_stack_switch(self, stack, *args):
         log.debug("Switched Stack! %r", args)
         #self.update_header_button()
 
     def on_send_stack_switch(self, stack, *args):
         log.debug("Switched Send Stack! %r", args)
-        #self.update_header_button()
+        current = self.send.stack.get_visible_child()
+        if current == self.send.klw:
+            log.debug("Key list page now visible")
+            self.on_keylist_mapped(self.send.klw)
+        elif current == self.send.kpw:
+            log.debug("Key present page now visible")
+            self.on_keypresent_mapped(self.send.kpw)
+        elif current == self.send.rb:
+            log.debug("Result page now visible")
+            self.on_resultbox_mapped(self.send.rb)
 
     def on_receive_stack_switch(self, stack, *args):
         log.debug("Switched Receive Stack! %r", args)
         #self.update_header_button()
 
-
     def on_send_header_button_clicked(self, button, *args):
-        # Here we assume that there is only one place where
+        # Here we assume that there is only two places where
         # we could have possibly pressed this button, i.e.
-        # from the keypresentwidget.
+        # from the keypresentwidget or the result page
         log.debug("Send Headerbutton %r clicked! %r", button, args)
+        current = self.send.stack.get_visible_child()
         klw = self.send.klw
-        self.send_stack.set_visible_child(klw)
-        
-        self.send.deactivate()
-
-
+        kpw = self.send.kpw
+        # If we are in the keypresentwidget
+        if current == kpw:
+            self.send_stack.set_visible_child(klw)
+            self.send.deactivate()
+        # Else we are in the result page
+        else:
```

Comment:
do you think it makes sense to explicitly check for this case and raise an error in the else branch otherwise?  I'm not sure whether the benefits are worth the hassle. I thought I'd just throw in that idea.